### PR TITLE
Add minimal distributed tracing deployment

### PR DESCRIPTION
**Pull request by pavolloffay**
Signed-off-by: Pavol Loffay <p.loffay@gmail.com>

This PR adds minimal tracing deployment, it is similar to the local example from https://github.com/observatorium/observatorium/pull/461.  The minimal configuration uses in-memory storage follow up PR will build on top of this PR and include persistent storage (Elasticsearch). 

This is my first PR and therefore be careful when reviewing. Especially I was not sure about the RBAC configuration. 

The goal of this PR is to deploy it into dev then staging environment. The tracing team is not familiar with the deployment process and any guidance would be much appreciated.

Open questions: 
- [x] The tracing deployment requires Jaeger and OpenTelemetry operator installed in the cluster. How should be these operators installed - they are installed programmatically via OLM subscriptions that are included in the generated manifest files.
- [x] Set namespace for tracing similar to `OBSERVATORIUM_LOGS_NAMESPACE` - it seems not needed right now
- [ ] Update observatorium jsonnet that correctly sets Jaeger exporter URLs in the OTEL collector and configures TLS ` ca_file: "/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt"`
- [ ] Update observatorium/api jsonnet that sets appprotocol to `h2c` https://github.com/observatorium/api/pull/301

### Testing

I have deployed the template to the RHOBS test cluster

```bash
make manifests
oc process -f resources/services/observatorium-traces-template.yaml  -p NAMESPACE=observatorium-traces-testing -o yaml | oc create -f -
 k edit deployment observatorium-observatorium-api -n observatorium-testing
 k edit cm  observatorium-observatorium-api   -n observatorium-testing -o yaml 
```

Get token:
```bash
./token-refresher --url=https://observatorium.apps.rhobs-testing.qqzf.p1.openshiftapps.com  --oidc.client-id=observatorium-rhobs-testing  --oidc.client-secret=f0795197-2cce-426d-9f42-foo  --log.level=debug --oidc.issuer-url=https://sso.redhat.com/auth/realms/redhat-external --oidc.audience=observatorium-telemeter-testing --file /tmp/token
cat /tmp/token
```

Send data via OTELcol
```yaml
# kubectl port-forward service/observatorium-xyz-observatorium-api 8090:8090 -n observatorium
# docker run --rm -it --net=host -v $PWD:/tmp/conf ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector:0.47.0 --config /tmp/conf/otelconf.yaml
extensions:

receivers:
  otlp:
    protocols:
      grpc:
      http:
  jaeger:
    protocols:
      thrift_binary:
      thrift_compact:
      thrift_http:
      grpc:

processors:

exporters:
  otlp:
    # gives  Unavailable desc = unexpected HTTP status code received from server: 502 (Bad Gateway); transport: received unexpected content-type \"text/html\""
#    endpoint: gprc-notls-observatorium-testing.apps.rhobs-testing.qqzf.p1.openshiftapps.com:80
    # gives Unavailable desc = unexpected HTTP status code received from server: 502 (Bad Gateway); transport: received unexpected content-type \"text/html\"" if insecure: false
#    endpoint: https://grpc-tls-edge-redirect-observatorium-testing.apps.rhobs-testing.qqzf.p1.openshiftapps.com:443
    endpoint: grpc-observatorium-testing.apps.rhobs-testing.qqzf.p1.openshiftapps.com:443
#    endpoint: localhost:8090
    compression: none
    headers:
#      Content-Type: application/grpc
      X-Tenant: "rhobs"
      Authorization: "bearer eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICItNGVsY19WZE5fV3NPVVlmMkc0UXhyOEdjd0l4X0t0WFVDaXRhdExLbEx3In0.eyJleHAiOjE2NTM5MTMwMTYsImlhdCI6MTY1MzkxMjExNiwianRpIjoiMTVkNTQ2NjItY2EzMy00MmM1LWEwOTYtOTU2ZTY2YmVlOTYxIiwiaXNzIjoiaHR0cHM6Ly9zc28ucmVkaGF0LmNvbS9hdXRoL3JlYWxtcy9yZWRoYXQtZXh0ZXJuYWwiLCJhdWQiOiJvYnNlcnZhdG9yaXVtLXRlbGVtZXRlci10ZXN0aW5nIiwic3ViIjoiNjRiODc4YTktNzYwZS00ODEyLWI5ZTQtOTFlYjA4MDFmMWM2IiwidHlwIjoiQmVhcmVyIiwiYXpwIjoib2JzZXJ2YXRvcml1bS1yaG9icy10ZXN0aW5nIiwiYWNyIjoiMSIsInJlYWxtX2FjY2VzcyI6eyJyb2xlcyI6WyJkZWZhdWx0LXJvbGVzLXJlZGhhdC1leHRlcm5hbCIsIm9mZmxpbmVfYWNjZXNzIiwidW1hX2F1dGhvcml6YXRpb24iXX0sInJlc291cmNlX2FjY2VzcyI6eyJyaGQtZG0iOnsicm9sZXMiOlsicmh1c2VyIl19LCJhY2NvdW50Ijp7InJvbGVzIjpbIm1hbmFnZS1hY2NvdW50IiwibWFuYWdlLWFjY291bnQtbGlua3MiLCJ2aWV3LXByb2ZpbGUiXX19LCJzY29wZSI6InByb2ZpbGUgZW1haWwiLCJjbGllbnRJZCI6Im9ic2VydmF0b3JpdW0tcmhvYnMtdGVzdGluZyIsImNsaWVudEhvc3QiOiI4MC4yMTkuNDQuMTM4IiwiZW1haWxfdmVyaWZpZWQiOmZhbHNlLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJzZXJ2aWNlLWFjY291bnQtb2JzZXJ2YXRvcml1bS1yaG9icy10ZXN0aW5nIiwiY2xpZW50QWRkcmVzcyI6IjgwLjIxOS40NC4xMzgiLCJlbWFpbCI6InRlYW0tbW9uaXRvcmluZ0ByZWRoYXQuY29tIn0.WwylQ_Lp3lsLis-bHF8VpRH6ENPmNxYq1y7an3H4lcMqy9iQ3_xRehSVwSxayCrSs4Gp8NQZRYSx3iF2mcu2kBqz9rmmFU7itSMCm_pGh7eoazNSbHEgN5t2_dDiShazB8im3MdDWtPDzma5UoARz0fWs9GzPGAE8MXdCi7TlnEJXHZujMkd7mSvHXn7AGppZDfJ6jYi0gRPGAbON_9vL9XQhL-n_bCHzdnRURWUAxvdL20AKBqqO_g0wrdS83GuDzfTmZYe264eYVLVTFqeLzBWEDNJZuAyvJHzaidrJRrEJb50jdtT3kEJCGal1qtPZ9tyfoyVBkGnA63LRMHCnyCn-Jr9K-_sVK49kktmZtu8gy-zUs-22FYG7aEpUEV4pqRESd9fvOGm1fmLv6rjuMjMlMsSt3kgjwjx4OAYttg8vUZsO13d6f0UI6Aw2gzRVol_IEpOKP-TivFMQyf6OpMUTDFycwoZjNsaHYKWghtNW1sqptayiav9GeuTWpMb4Q1Vkf0uKhFau_pYrJ09QAX-bB5t6HiAgCd14vKDatitJT2GfR71O62udhcRNaPCtl-MfDTcyiRLg7wOE8SYXfboNonBibUBfARyAnn1ynF7GyJwQqjAwPHFBQ4S47lKaICr-i1tZxfeDE9BzjEpoxATWr9oXmHxpzyXyiIxhgU"
    tls:
      insecure: true

service:
  pipelines:
    traces:
      receivers: [otlp, jaeger]
      exporters: [otlp]
```

Test with grpcurl
```bash
docker run --net=host fullstorydev/grpcurl -plaintext -vv gprc-notls-observatorium-testing.apps.rhobs-testing.qqzf.p1.openshiftapps.com:80 list
```

### CODE REVIEW

File: resources/services/observatorium-template.yaml [hunk 1]
```diff
@@ -1050,6 +1052,74 @@ objects:
         app.kubernetes.io/instance: observatorium
         app.kubernetes.io/name: observatorium-up
         app.kubernetes.io/part-of: observatorium
+- apiVersion: jaegertracing.io/v1
```
Comment (pavolloffay) [inline resources/services/observatorium-template.yaml:1055]: Am I right that the tracing deployment should be created in `observatorium-traces-template.yaml` and not in this file?

Comment (bwplotka) [reply resources/services/observatorium-template.yaml:1057]: Yes, this allows us to deploy those things in separation (:

**Review by bwplotka — COMMENTED** [review]
Nice, this makes sense, just indeed let's create a separate file for traces like we have for logs and metrics. 👍🏽

File: resources/services/observatorium-template.yaml [hunk 2]
```diff
@@ -1050,6 +1054,74 @@ objects:
...
+        pipelines:
+          traces:
+            exporters:
+            - "jaeger/rhobs"
+            - "jaeger/telemeter"
+            processors:
+            - "routing"
+            receivers:
+            - "otlp"
+    image: ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector-contrib:0.46.0
```
Comment (PhilipGough) [inline resources/services/observatorium-template.yaml:1123]: @bwplotka will this be ok? I know we had some issues (in terms of app-interface not being happy) when we tried to roll out parca using the upstream `ghcr` image

**Comment by pavolloffay** [comment]
@bill3tt  @bwplotka could you please comment on the open question on how to install the operator? 

> The tracing deployment requires Jaeger and OpenTelemetry operator installed in the cluster. How should be these operators installed?

File: resources/services/observatorium-template.yaml (hunk 1 above)
Comment (pavolloffay) [reply resources/services/observatorium-template.yaml]: done, the tracing objects were moved to a separate manifest file

**Comment by esnible** [comment]
I'd like to test this PR against my team's OpenShift cluster.

I see the file _tests/launch.sh_ in this repo but it doesn't seem to have the same settings I see on the rhobs-testing dev cluster.  Where is the script that deploys Observatorium on the dev cluster?

**Review by periklis — APPROVED** [review]
LGTM

File: resources/services/observatorium-traces-template.yaml [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+- name: OPENTELEMETRY_OPERATOR_VERSION
+  value: 0.44.1-2
+- name: OPENTELEMETRY_OPERATOR_NAMESPACE
+  value: openshift-operators
+- name: OPENTELEMETRY_OPERATOR_SOURCE
+  value: redhat-operators
+- name: JAEGER_OPERATOR_VERSION
+  value: 1.30.2
+- name: JAEGER_OPERATOR_NAMESPACE
+  value: openshift-operators
```
Comment (pavolloffay) [inline resources/services/observatorium-traces-template.yaml:110]: @rubenvp8510 in which ns the jaeger-operator should be installed?

File: resources/services/observatorium-template.yaml [hunk 4]
```diff
@@ -246,6 +248,8 @@ objects:
           - --logs.read.endpoint=http://observatorium-loki-query-frontend-http.${OBSERVATORIUM_LOGS_NAMESPACE}.svc.cluster.local:3100
           - --logs.tail.endpoint=http://observatorium-loki-querier-http.${OBSERVATORIUM_LOGS_NAMESPACE}.svc.cluster.local:3100
           - --logs.write.endpoint=http://observatorium-loki-distributor-http.${OBSERVATORIUM_LOGS_NAMESPACE}.svc.cluster.local:3100
+          - --traces.write.endpoint=observatorium-otel-collector:4318
```
Comment (bill3tt) [inline resources/services/observatorium-template.yaml:251]: Currently - this does not point to a name-spaced service. I expect the OTEL collector will be behind a service like `observatorium-otel-collector.observatorium-tracing-testing.svc.cluster.local`

Comment (pavolloffay) [reply resources/services/observatorium-template.yaml]: Fixed, the URL is parametrized now

File: resources/services/observatorium-traces-template.yaml (hunk 3 above)
Comment (rubenvp8510) [reply resources/services/observatorium-traces-template.yaml:113]: it is deployed on `openshift-distributed-tracing`

File: resources/services/observatorium-traces-template.yaml [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+      app.kubernetes.io/instance: observatorium
+      app.kubernetes.io/name: otelcol
+      app.kubernetes.io/part-of: observatorium
+    name: observatorium-otel
+    namespace: ${NAMESPACE}
+  spec:
+    config: |-
+      exporters:
+        jaeger/rhobs:
+          endpoint: "observatorium-jaeger-rhobs-collector:14250"
```
Comment (pavolloffay) [inline resources/services/observatorium-traces-template.yaml:68]: This need to be changed after https://github.com/observatorium/observatorium/pull/485 is merged

**Comment by pavolloffay** [comment]
Bumped obs/obs/configuration jsonnet by ` jb update https://github.com/observatorium/observatorium/configuration@main` which bumped more versions.

**Comment by pavolloffay** [comment]
I would like to get this merged. Merging this should not have affect on the app-sre configuration. The `grpc-public` is not exposed outside the cluster

File: resources/services/observatorium-traces-template.yaml (hunk 5 above)
Comment (douglascamata) [reply resources/services/observatorium-traces-template.yaml:68]: Is this already changed, @pavolloffay? https://github.com/observatorium/observatorium/pull/485 is merged.

Comment (pavolloffay) [reply resources/services/observatorium-traces-template.yaml:68]: https://github.com/rhobs/configuration/pull/223 should go in first before I can change it here

**Review by bwplotka — APPROVED** [review]
Thanks! LGTM 👍🏽

**Pull request merged**